sqlalchemy-stubs/sql/sqltypes.pyi

```diff
@@ -3,78 +3,107 @@ from datetime import datetime
-    RETURNS_UNICODE: Any = ...
-    RETURNS_BYTES: Any = ...
-    RETURNS_CONDITIONAL: Any = ...
-    RETURNS_UNKNOWN: Any = ...
-    length: Any = ...
-    collation: Any = ...
+_TE = TypeVar("_TE", bound=TypeEngine[Any])
+_ST = TypeVar("_ST", bound=SchemaType)
+_EN = TypeVar("_EN", bound=Enum)
+
+_NumericType = Union[float, Decimal]
+_JSONType = Union[str, Mapping[Any, Any], List[Any]]
+
+class _LookupExpressionAdapter(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]): ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class Concatenable(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]): ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class Indexable(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]):
+        def __getitem__(self, index: Any) -> Any: ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class String(Concatenable[str], TypeEngine[str]):  # type: ignore[misc]
+    __visit_name__: str = ...
+    RETURNS_UNICODE: ClassVar[langhelpers._symbol] = ...
+    RETURNS_BYTES: ClassVar[langhelpers._symbol] = ...
+    RETURNS_CONDITIONAL: ClassVar[langhelpers._symbol] = ...
+    RETURNS_UNKNOWN: ClassVar[langhelpers._symbol] = ...
+    length: Optional[int] = ...
+    collation: Optional[str] = ...
+    @overload
     def __init__(
         self,
-        length: Optional[Any] = ...,
-        collation: Optional[Any] = ...,
+        length: Optional[int] = ...,
+        collation: Optional[str] = ...,
         convert_unicode: bool = ...,
-        unicode_error: Optional[Any] = ...,
         _warn_on_bytestring: bool = ...,
         _expect_unicode: bool = ...,
     ) -> None: ...
-    def literal_processor(self, dialect: Any): ...
-    def bind_processor(self, dialect: Any): ...
-    def result_processor(self, dialect: Any, coltype: Any): ...
+    @overload
+    def __init__(
+        self,
+        length: Optional[int],
+        collation: Optional[str],
+        convert_unicode: Literal["force"],
+        unicode_error: str,
+        _warn_on_bytestring: bool = ...,
+        _expect_unicode: bool = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        length: Optional[int],
+        *,
+        convert_unicode: Literal["force"],
+        unicode_error: str,
+        _warn_on_bytestring: bool = ...,
+        _expect_unicode: bool = ...,
+    ) -> None: ...
+    def literal_processor(
+        self, dialect: Dialect
+    ) -> _LiteralProcessor[str]: ...
+    def bind_processor(
+        self, dialect: Dialect
+    ) -> Optional[_BindProcessor[str]]: ...
+    def result_processor(
+        self, dialect: Dialect, coltype: Any
+    ) -> Optional[_ResultProcessor[str]]: ...
     @property
-    def python_type(self): ...
-    def get_dbapi_type(self, dbapi: Any): ...
+    def python_type(self) -> Union[Type[str], Type[util.text_type]]: ...
+    def get_dbapi_type(self, dbapi: Any) -> Any: ...
```

Comment:
I noticed that we use Any here instead of the actual type (`Type[STRING]` I think) Does it has a particular reason? Any is fine by me since this is not really a public method

Reply:
I figured we can go back through and get more specific after this initial pass is done

Reply:
ok, sure